runtime/JavaScript/src/antlr4/tree/ParseTreeVisitor.d.ts

```diff
@@ -1,16 +1,18 @@
-import {RuleNode} from "./RuleNode";
-import {ErrorNode} from "./ErrorNode";
-import {TerminalNode} from "./TerminalNode";
-import {ParseTree} from "./ParseTree";
+/* Copyright (c) 2012-2022 The ANTLR Project. All rights reserved.
+ * Use of this file is governed by the BSD 3-clause license that
+ * can be found in the LICENSE.txt file in the project root.
+ */
 
-export declare class ParseTreeVisitor<Result> {
-
-    visit(tree: ParseTree): Result;
-
-    visitChildren(node: RuleNode): Result;
+import ErrorNode from "./ErrorNode.js";
+import TerminalNode from "./TerminalNode.js";
+import ParseTree from "./ParseTree.js";
 
-    visitTerminal(node: TerminalNode): Result;
-
-    visitErrorNode(node: ErrorNode): Result;
+export declare class ParseTreeVisitor<Result> {
+    public visit(tree: ParseTree): Result;
+    public visitChildren(node: ParseTree): Result;
```

Comment:
a node with children is necessarily a ruleNode, please revert

Reply:
No, a node with children is a (Tree)[https://github.com/antlr/antlr4/blob/dev/runtime/Java/src/org/antlr/v4/runtime/tree/Tree.java], but that's an interface in Java and there's no real sense to separate the 3 interfaces `Tree`, `SyntaxTree` and `ParseTree`. The first two don't add anything useful to `ParseTree` hence they are combined into a single interface.

The `RuleNode` is another interface which only adds a single member. There's only one type which uses `RuleNode`, which is `RuleContext`, so it makes sense to remove `RuleNode` and directly work with `RuleContext`. Why keep a class/interface which is not being used?

Reply:
"it makes sense" is an opinion. We just have different coding styles, and I don't change working code to satisfy coding style, otherwise I would change the code 10 times per day.

Reply:
Then read "make sense" as "there's a good reason". It's not personal coding style, but something useful. You also have left out things in the current runtime that you have no use for. Btw. I did the same change in the C++ runtime in 2016. The 3 different interfaces are simply over-engineered and are useless in the context of the JS/TS runtime.

Reply:
I happen to find it more useful to show that the 3 types of nodes are covered by the if/else if... which is why I wrote it like that in the first place.

Reply:
Can you point me to a place in source code where you test, say, on `Tree`? How often do you check for `ParseTree` and for `SyntaxTree`? I cannot find any place where `Tree` or `SyntaxTree` is used in JS/TS code, except to define `ParseTree`.

And even in the other targets there's only very few use of `Tree`, namely in the `Trees` file (only Java, Dart and Swift) and in the Java UI tool, which is not relevant here.

I think these are enough points to warrant a simplification of this small corner of the ANTLR4 type tree. That's why I have no intention to revert that.